Allow chesster to update schedules for us

lakinwecker opened the pull request:
This is working. I haven't had a chance to do a second pass reviewing the code myself (which I'd like to do), but I figured I would get it into your hands sooner than later.

It requires a bunch of new settings, like these:

```
{
    "token":"xoxb-31643488983-3DW3H5kYafnzdULXLPikFmyY",
    "serviceAccountAuth": {
        "client_email": "chesster@chesster-lichess-4545-bot.iam.gserviceaccount.com",
        "private_key": " .... "
    },
    "scheduling": {
        "unstable_bot-lonewolf": {
            "key": "...",
            "colname": "gamescheduledingmt",
            "offset_hours": 1,
            "format": "MM/DD HH:mm"
        },
        "unstable_bot": {
            "key": "...",
            "colname": "timemmddhhmm",
            "offset_hours": 0,
            "format": "MM/DD @ HH:mm"
        }
    }
}
```

"key" is the spreadsheet "key. and "private_key" is the private key from a service account.  I have one setup and pointed at my testing spreadsheets. You will need to do the same if you want to test this out, or I can spin it up for you.

The client_email is the service account email and the spreadsheet needs to be shared with that email. Also, you will have to unprotect the range you want it to edit.  colname is a lower case + special chars removed version of the column you want it to edit, and 'format' is the format of the date you want it to put into that column.

endrawes0 commented on scheduling.js:
```diff
@@ -0,0 +1,202 @@
+    "fri",
+    "sat",
+
+    // TODO: yes, this is oddly specific - but unless he changes his username ....
```
Ha. What?

lakinwecker replied:
greentiger (on lichess) uses the name steveharwell on slack. This helps deal with that. He's not in lone wolf, or I would have asked him to change his slack username.

endrawes0 commented:
I like the moment API. That simplifies a lot. 

Look forward to doing the full review after refactoring. Thanks again.

lakinwecker commented:
Yeah, moment is a great library.

lakinwecker commented:
Alright, the refactoring I wanted to do before review is done.  The code isn't perfect, but it works.

The primary issue I see right now, is that in an effort to accept all sorts of date formats I have chesster guess at the date and run with it if it's in the appropriate range for the round.  However, this results in chesster recognizing gibberish as dates and happily updating the spreadsheet.

I'm torn between removing his ability to parse more formats but have our confidence in the data be better, or leaving it in.  At the moment, I'm leaning towards leaving in. For three reasons:

1. The primary problem is when people literally put in gibberish, like lakinwecker vs endrawes0 asdfasdf asd fas asdf  which will be parsed.  If people put this in to trick the bot, then it's their own fault.
2. You have to put in the appropriate pairing to go with the gibberish. 
3. Chesster immediately responds with his version of the schedule. People will be able to check it at that point and confirm it's correct. If they don't and they used gibberish - that's the players fault. IMO.

endrawes0 commented on chesster.js:
```diff
@@ -1169,3 +1177,71 @@ controller.hears([
+            var results = spreadsheets.parse_scheduling(message.text, {
+                offset_hours: scheduling_options.offset_hours
+            });
+            var white = users.getByNameOrID(results.white);
```
This operation seems unnecessary. if we only want the name, then can we just return the name, or rather, can we just access the nae in update_schedule?

Not that important, but simplifies this slightly.

lakinwecker replied:
I know you said to ignore this from below, but the other reason we do this, is that when the @ version of names are used, we don't get their username. We get their id. Literally if I say:

@endraws vs @lakinwecker

the message looks like:
@<Uds7sD)7fHS> vs @<UHSDHF728>  Where the angle brackets surround a user id

endrawes0 replied:
Interesting. That's the 'markup' form. I guess it makes sense.
On Apr 17, 2016 9:58 AM, lakinwecker <notifications@github.com> wrote:In chesster.js:
> +
> +/* Scheduling */
> +
> +// reply to any incoming message
> +controller.on('ambient', function(bot, message) {
> +    bot_exception_handler(bot, message, function(){
> +        var channel = channels.byId[message.channel];
> +        var scheduling_options = config.scheduling[channel.name];
> +        if (!scheduling_options) {
> +            return;
> +        } 
> +        try {
> +            var results = spreadsheets.parse_scheduling(message.text, {
> +                offset_hours: scheduling_options.offset_hours
> +            });
> +            var white = users.getByNameOrID(results.white);

I know you said to ignore this from below, but the other reason we do this, is that when the @ version of names are used, we don't get their username. We get their id. Literally if I say:

@endraws vs @lakinwecker

the message looks like:
@ vs @  Where the angle brackets surround a user id

—You are receiving this because you commented.Reply to this email directly or view it on GitHub

endrawes0 commented on chesster.js:
```diff
@@ -1169,3 +1177,71 @@ controller.hears([
+                            white = black;
+                            black = tmp;
+                        }
+                        var whiteDate = results.date.clone().utcOffset(white.tz_offset/60);
```
I really like this. I think it will make things more clear for BOTH parties.

I see why you need white/black above


Ignore my previous comment

endrawes0 commented on spreadsheets.js:
```diff
@@ -0,0 +1,215 @@
+    "sat",
+
+    // TODO: yes, this is oddly specific - but unless he changes his username ....
+    "steveharwell"
```
I see. I'll ask steve about his name. perhaps he would prefer greentiger.

lakinwecker replied:
For what it's worth, it's extremely easy to change your username with slack.  I would prefer he used his lichess username on slack.  It makes loads of stuff easier.

endrawes0 commented on spreadsheets.js:
```diff
@@ -0,0 +1,215 @@
+
+    // Now build up some possible strings and try a bunch of patterns
+    // to find a date in our range.
+    var extrema = get_round_extrema(options);
```
please confirm my understanding here:

essentially, this will match if its

`[ignorable strings] @<opponent1> [ignorable strings] @<opponent2> [ignorable strings] <datestring> [ignorable strings]`

yeah?

lakinwecker replied:
You are correct.  It ignores the the ignorable words from any location in the message.  It's a bit heavy handed, but it works in most cases.

endrawes0 commented on spreadsheets.js:
```diff
@@ -0,0 +1,215 @@
+    }
+    round_start.hour(0).minute(0).second(0);
+    round_end = round_start.clone().add(7, 'days');
+    if (options.offset_hours) {
```
whats happening here? I dont understand what offset_hours is for? because we pair at a different time?

lakinwecker replied:
I will document this method a bit better. 

To answer your question:

Yes, this deals with the 1 hour offset from the bounds on the round between team and lonewolf.

endrawes0 commented on spreadsheets.js:
```diff
@@ -0,0 +1,215 @@
+            return callback(err);
+        }
+        // This portion is specific to the team tournament
+        row[colname] = date.format(format);
```
cell by column name is nice. glad this is part of the api. much better than my hard coded indices. ill have to remove them and rewrite my stuff to use the row interface where possible.

endrawes0 replied:
and add configuration for column names. great improvement.

endrawes0 commented on test/test_spreadsheets.js:
```diff
@@ -0,0 +1,337 @@
+    });
+});
+
+
```
how do I run the unit tests?

endrawes0 replied:
node ./test/<test>.json

lakinwecker replied:
there are new dependencies. I will update the README to include them. I put them in the vagrant-dependencies, but I think we should have an npm-dependencies script that we can use as well. I'll do both today.

Once the new dependencies are installed there is a "mocha" command that runs the tests for me.

endrawes0 commented on chesster.js:
```diff
@@ -3,6 +3,7 @@ var Botkit = require('botkit');
 var fuzzy = require('./fuzzy_match.js');
+var spreadsheets = require('./spreadsheets.js');
 
 var SHEET_URL = "https://lichess4545.slack.com/files/mrlegilimens/F0VNACY64/lichess4545season3-graphs";
```
these should be configurable as well. Ill work on that eventually as well. I need to find some time to do some programming.

lakinwecker commented:
I've made a number of improvements:

1. I've refactored the configuration to be more clear and flexible. This includes  a "warning" period, where we will accept your scheduled time but warn you that you are cutting it close. 
2. I've refactored the responses out of the scheduling method to make it a bit easier to read.
3. I've added protection against updating other people's schedules.
4. I've added better messaging to let people know why their schedule is not accepted.
5. I've fixed some bugs with the schedule parsing.  We are going to have to be careful here as this code combined with moment is very forgiving on the date formats, and if people don't double check the time that chesster reports back - we might end up with some strange times in the spreadsheet.

Anyways, I'm done with this for about a week. In other words, aside from very small fixes I doubt I'll have time to work on it again till next weekend. So please feel free to make your own improvements/changes/fixes to it as you see fit.

endrawes0 commented on chesster.js:
```diff
@@ -1169,3 +1177,127 @@ controller.hears([
+// Can't parse the date
+function scheduling_reply_cant_parse(bot, message) {
+    var user = "<@"+message.user+">";
+    bot.reply(message, ":x: " + user + " I don't understand. Please use a format like: @lakinwecker v @lakinwecker 04/16 @ 16:00 GMT");
```
I'm probably being silly, but this may be misleading.

I'll probably change this to say: @white v @black 04/16 @ 16:00

endrawes0 commented on chesster.js:
```diff
@@ -1169,3 +1177,127 @@ controller.hears([
+// can't find the users you menteiond
+function scheduling_reply_cant_find_user(bot, message) {
+    var user = "<@"+message.user+">";
+    bot.reply(message, ":x: " + user + " I don't recognize one of the players you mentioned.");
```
Can we use the fuzzy match to determine the user?

endrawes0 commented:
I'm accepting the pull request. 

I won't be able to update Chesster until this evening. If you get a chance to do it yourself, feel free.

Other wise, enjoy your week. Thanks for the great contribution.
